test/hotspot/gtest/utilities/test_resourceHash.cpp

```diff
@@ -287,3 +288,110 @@ TEST_VM_F(GenericResourceHashtableTest, bad_hash_no_rm) {
 TEST_VM_F(GenericResourceHashtableTest, identity_hash_no_rm) {
   Runner<identity_hash, primitive_equals<K>, 1, ResourceObj::C_HEAP>::test(512);
 }
+
+class ResourceHashtableDeleteTest : public ::testing::Test {
+ public:
+    class TestValue : public CHeapObj<mtTest> {
+        Symbol* _s;
+      public:
+        // Never have ctors and dtors fix refcounts without copy ctors and assignment operators!
+        // Unless it's declared and used as a CHeapObj with
+        // NONCOPYABLE(TestValue)
+        TestValue(Symbol* name) : _s(name) { _s->increment_refcount(); }
+        TestValue(const TestValue& tv) { _s = tv.s(); _s->increment_refcount(); }
+        TestValue& operator=(const TestValue& tv) { _s = tv.s(); _s->increment_refcount(); return *this; }
+        ~TestValue() { _s->decrement_refcount(); }
+        Symbol* s() const { return _s; }
+    };
+
+    // ResourceHashtable whose value is a copy of TestValue.
+    ResourceHashtable<Symbol*, TestValue, 107, ResourceObj::C_HEAP, mtTest> _test_table;
+
+    class Deleter : public StackObj {
+      public:
+        bool do_entry(Symbol*& key, TestValue& value) {
+          // decrement the refcount in key or not?
+          // Since we didn't increment the key, we shouldn't decrement it.
+          // Calling delete on the hashtable Node which contains value will
+          // decrement the refcount.  That's actually best since the whole
+          // entry will be gone at once.
+          return true;
+        }
+    };
+
+    // ResourceHashtable whose value is a pointer to TestValue.
+    ResourceHashtable<Symbol*, TestValue*, 107, ResourceObj::C_HEAP, mtTest> _ptr_test_table;
+
+    class PtrDeleter : public StackObj {
+      public:
+        bool do_entry(Symbol*& key, TestValue*& value) {
+          // decrement the refcount in key or not?
+          // Since we incremented the key, in this case, we should decrement it.
+          key->decrement_refcount();
+          // If the hashtable value is a pointer, need to delete it from here.
+          delete value;
+          // The risk is that if this is the last reference to the key, it'll be
+          // pointed to in the table with refcount 0.
+          // But the node is going to be deleted next anyway, so a pointer to dead data
+          // is ok?
+          return true;
+        }
+    };
+
+};
+
+TEST_VM_F(ResourceHashtableDeleteTest, check_delete) {
+  TempNewSymbol s = SymbolTable::new_symbol("abc");
+  int s_orig_count = s->refcount();
+  {
+    TestValue tv(s);
+    // If you use a Symbol* as key in the ResourceHashtable, you must increment the
+    // refcount outside the hashtable functions.
+    s->increment_refcount();
+    _test_table.put(s, tv);
+    ASSERT_EQ(s->refcount(), s_orig_count + 3) << "refcount incremented";
+  }
+  ASSERT_EQ(s->refcount(), s_orig_count + 2) << "refcount not copied";
```

Comment:
I may misunderstand things, but why do you test the same sequence twice? The first one above seems not to match your description since the value holds the Symbol secure during its lifetime.

If you wanted to test what the comment describes you would add a value that does nothing to the Symbol*. But I'm not sure that is even worth testing since it is clear what happens (if you put Symbol* as key and whatever as value, nobody will increase the refcount of Symbol*).

Reply:
Yes, the first line (344) holds the value through the lifetime but I assert that at the end the refcount is where we started.
The second sequence tests the unlink function, which acts differently than remove. In remove, we have to do any refcounting cleanup for the key except destruction because remove doesn't have a callback.  In unlink, we have to do the cleanup in the callback.
I'm not sure what your suggested test would do, except underflow the refcount of the symbol, which I don't want to do because then it'd be a negative test.